File: src/main/java/agentcore/agentcommunication/AgentCommunicationImpl.java
package agentcore.agentcommunication;

import java.io.*;
import java.net.Socket;
import java.util.Observable;

import static java.lang.Thread.interrupted;

/**
 * Created by user on 21.02.2017.
 */
public class AgentCommunicationImpl extends Observable implements IAgentCommunication, Runnable {

    private ObjectOutputStream outputStream;
    private ObjectInputStream inputStream;
    private Socket socket;
    private boolean isConnect = false;
    private Thread thread;

    public AgentCommunicationImpl() { }

    @Override
    public void sendMassege(AMessage message) throws IOException {
        if (!isConnect)
            return;

        outputStream.writeObject(message);
        outputStream.flush();
    }

    @Override
    public void connect(String host, int port) throws IOException {
        if (isConnect)
            return;

        // создаём подключение
        try {
            socket = new Socket(host, port);
        } catch (IOException e) {
            System.out.println(e.toString());
            return;
        }

        // получаем потоки ввода вывода
        try {
            InputStream in = socket.getInputStream();
            OutputStream out = socket.getOutputStream();
            outputStream = new ObjectOutputStream(out);
            inputStream = new ObjectInputStream(in);
        } catch (Exception e) {
            throw new IOException(e.toString() + " Не удалось получить поток вывода.");
        }

        isConnect = true;
        startThread();
    }

    @Override
    public void disconnect() throws IOException {
        if (!isConnect)
            return;

        try {
            isConnect = false;
            thread.interrupt();
            outputStream.close();
            inputStream.close();
            socket.close();
        } catch (IOException e) {
            throw new IOException(e.toString());
        }
    }

    @Override
    public boolean isConnect() {
        return isConnect;
    }

    /**
     * Получение сообщений от сервера
     */
    @Override
    public void run() {
        // наследоваться от потока + сделать чтение нормальным(чтобы читать данные, когда они есть)
        try {
            while (!interrupted()) {
                Object object = inputStream.readObject();
                checkServerObject(object);
            }
        } catch (Exception e) {
            System.out.println(e.toString() + " ошибка чтения данных с сервера");
        }
    }

    /**
     * Проверяет и записывает данные о пришедшем сообщении с сервера
     * @param object присланный объект
     */
    private void checkServerObject(Object object) {
        if (object instanceof AMessage) {
            System.out.println("Пришло сообщение с сервера AgentCommunicationImpl");
            setChanged();
            notifyObservers(object);
        }
    }

    /**
     * запускает выполнение в новом потоке
     */
    private void startThread() {
        thread = new Thread(this);
        thread.setDaemon(true);
        thread.start();
    }

}

File: agentcore/src/main/java/agentcore/agentcommunication/AgentCommunicationImpl.kt
package agentcore.agentcommunication

import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.net.Socket
import java.util.*

import java.lang.Thread.interrupted

/**
 * @author Nikita Gorodilov
 */
class AgentCommunicationImpl : Observable(), IAgentCommunication, Runnable {

    private var outputStream: ObjectOutputStream? = null
    private var inputStream: ObjectInputStream? = null
    private var socket: Socket? = null
    private var isConnect = false
    private var thread: Thread? = null

    @Throws(IOException::class)
    override fun sendMassege(message: Message) {
        if (!isConnect)
            return

        outputStream!!.writeObject(message)
        outputStream!!.flush()
    }

    @Throws(IOException::class)
    override fun connect(host: String, port: Int) {
        if (isConnect)
            return

        // создаём подключение
        try {
            socket = Socket(host, port)
        } catch (e: IOException) {
            println(e.toString())
            return
        }

        // получаем потоки ввода вывода
        try {
            val input = socket!!.getInputStream()
            val out = socket!!.getOutputStream()
            outputStream = ObjectOutputStream(out)
            inputStream = ObjectInputStream(input)
        } catch (e: Exception) {
            throw IOException(e.toString() + " Не удалось получить поток вывода.")
        }

        isConnect = true
        startThread()
    }

    @Throws(IOException::class)
    override fun disconnect() {
        if (!isConnect)
            return

        try {
            isConnect = false
            thread!!.interrupt()
            outputStream!!.close()
            inputStream!!.close()
            socket!!.close()
        } catch (e: IOException) {
            throw IOException(e.toString())
        }
    }

    override fun isConnect(): Boolean {
        return isConnect
    }

    /**
     * Получение сообщений от сервера
     */
    override fun run() {
        // наследоваться от потока + сделать чтение нормальным(чтобы читать данные, когда они есть)
        try {
            while (!interrupted()) {
                val obj = inputStream!!.readObject()
                checkServerObject(obj)
            }
        } catch (e: Exception) {
            println(e.toString() + " ошибка чтения данных с сервера")
        }
    }

    /**
     * Проверяет и записывает данные о пришедшем сообщении с сервера
     * @param object присланный объект
     */
    private fun checkServerObject(obj: Any) {
        if (obj is Message) {
            println("Пришло сообщение с сервера AgentCommunicationImpl")
            setChanged()
            notifyObservers(obj)
        }
    }

    /**
     * запускает выполнение в новом потоке
     */
    private fun startThread() {
        thread = Thread(this)
        thread!!.isDaemon = true
        thread!!.start()
    }
}
